core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -130,6 +130,16 @@ private[spark] class Executor(
   private val urlClassLoader = createClassLoader()
   private val replClassLoader = addReplClassLoaderIfNeeded(urlClassLoader)
 
+  // One thread will handle loading all of the plugins on this executor
```

Comment:
Why a thread? This risks having the executor start doing things before plugins have had a chance of properly starting up.

Reply:
Originally was all in one thread, changed to separate based on this discussion: https://github.com/apache/spark/pull/21923#discussion_r206905166

I see the merits of both arguments, can change it based on the general consensus.

Reply:
@tgravescs thoughts?

Reply:
I guess it does depend on what the intended use is here.  If we have it in the same thread it has the issue that it could block the executor or take to long and things start timing out.  It can have more direct impact on the executor code itself, where as a separate thread isolates it more.   But like you say if its not here and we don't wait for it then we could have order issue if certain plugins have to be initialized before other things happen.  I can see both arguments as well.    So perhaps the api needs an init type function that can be called more inline with a timeout to prevent from taking to long and the main part of the plugin called in a separate thread?

Reply:
I suggested an `init` method in another comment already, and I think that should be enough, if coupled with documentation saying that method is calling inline with the initialization of the executor, and should avoid blocking it, doing any expensive initialization asynchronously.

We could try to be fancy and add both "init()" and "initAsync" or something to make it easier for plugin writers, but that sounds like a small gain for the extra complexity in the interface.

Reply:
Aside from semantics, would an `init` method be necessary instead of having the initialization logic be in the plugin's constructor? Since the class loader is going to call the constructor immediately, I figure having an `init` function would only really make a difference if we want to load the plugins right here, and then call `init` at a later point in the executor's creation. I can't think of any particular reason why we'd want to do that, unless there's specific executor structures that we want created prior to plugin initialization (although in that case we could also just move the plugin initialization further down)

Reply:
I guess it could be in the constructor; `Utils.loadExtensions` already provides a `SparkConf` to the constructor if one accepts it, which was the only thing I could think of.

I generally dislike plugin APIs that encourage initialization in the constructor, but here, other than maybe potentially some benefit for testing, I'm not seeing a lot of differences in not having the init method after all...

Reply:
That does make sense. While I did say "aside from semantics", semantics is a good reason to include it. Especially since it'll be harder to get plugin writers to adopt an `init` function later. I'll make the other changes and make sure the tests still pass, if anyone does feel strongly (or even weakly) on one way over another I don't think there's much harm in either approach.